exonum-java-binding/testkit/src/main/java/com/exonum/binding/testkit/TestKit.java

```diff
@@ -493,11 +477,36 @@ public Builder withTimeService(TimeProvider timeProvider) {
     public TestKit build() {
       checkCorrectServiceNumber(services.size());
       checkCorrectValidatorNumber();
-      return newInstance(services, nodeType, validatorCount, timeProvider);
+      List<TestKitServiceInstances> testKitServiceInstances = mergeServiceSpecs();
+      return newInstance(testKitServiceInstances, nodeType, validatorCount, timeServiceSpec);
+    }
+
+    /**
+     * Turn collection of service instances into a list of
+     * {@linkplain TestKitServiceInstances} objects for native to work with.
+     */
+    private List<TestKitServiceInstances> mergeServiceSpecs() {
+      Set<ServiceArtifactId> serviceArtifactIds = services.keySet();
+      checkArgument(serviceArtifactIds.containsAll(serviceArtifactFilenames.keySet()),
+          "All service instances that are deployed should also be instantiated"
+              + " and vice versa.");
+      return serviceArtifactIds.stream()
+          .map(this::aggregateServiceSpecs)
+          .collect(toList());
+    }
+
+    /**
+     * Aggregates service instances specifications of a given service artifact id as a
+     * {@linkplain TestKitServiceInstances} object.
+     */
+    private TestKitServiceInstances aggregateServiceSpecs(ServiceArtifactId artifactId) {
+      String artifactFilename = serviceArtifactFilenames.get(artifactId);
+      ServiceSpec[] serviceSpecs = services.get(artifactId).toArray(new ServiceSpec[0]);
+      return new TestKitServiceInstances(artifactId.toString(), artifactFilename, serviceSpecs);
     }
 
     private void checkCorrectValidatorNumber() {
```

Comment:
Actually, on `checkCorrectServiceNumber`: 

Given the dynamic nature of the runtime and the fact that it's not as easy
       to register a service just to make testkit happy when that service might not be needed,
       I'd reconsider 0.

Reply:
But why would a user want to create a TestKit with no custom services instantiated? There is currently no way to create services after building TestKit, as we do not expose the runtime, right?

Reply:
Yes, one won't be able to deploy and start new services. A user-service developer probably won't, I considered our cases when no service is needed for testing purposes (e.g., tests of time service support in Java).

Reply:
> Yes, one won't be able to deploy and start new services.

They can probably use deploy and start transactions if there is a single validator, but that does not appear easy enough to be useful.